### ISSUE TITLE
Stop pinning dependencies

### ISSUE THREAD

**Issue by ssbarnea** [python-jsonschema/check-jsonschema#259]
Apparently https://github.com/python-jsonschema/check-jsonschema/blob/main/setup.cfg#L22 this tool is pinning dependencies which can create conflicts on installations.

To be more precise, the pinning `ruamel.yaml==0.17.21` makes no sense as it will prevent installation of newer bugfix version such `0.17.22`.

Basically this pinning is a time bomb, one that increases maintenance work on this project.

If there is a desire to declare which versions of the dependencies were tested at some point, you could use a `lock` extra which does pin the version, but not the default dependencies.

Users may decide to install either `check-jsonschema[lock]` or just `check-jsonschema` based on their needs. Now they don't have an option.

I faced this [installation failure](https://github.com/ansible/ansible-lint/actions/runs/4863745302/jobs/8671874881?pr=3392) when I had to require the newer version of ruamel.yaml because it contained a bugfix that we needed.

**Comment by sirosen** [comment]
Given that this is an application, not a library, users should have no expectation that installing into the same virtualenv as other applications is going to work. I think pinning is pretty appropriate for applications (and inappropriate for libraries, though my views towards semver are a bit more generous than e.g. the pallets group).
I can have my mind changed on this, but coming in hot by saying that this is a "time bomb", in concert with installation into a n environment which apparently includes other tools, doesn't make for a very convincing argument yet.

I don't think removing pins for dependencies is appropriate unless they have a clear backwards compatibility contract. `ruamel.yaml` is pretty fuzzy about what we can expect for any given version. Maybe it should be `<0.18.0` based on the docs, but I wouldn't want the YAML customizations here to break for a user simply because the pinning became loose.

For now I'm updating ruamel.yaml to the latest -- which is hot off the presses and it isn't even in that project's changelog yet. I plan to put out a release when I'm able to come back to and polish off the work I had in progress.

**Comment by sirosen** [comment]
With 0.23.0 released, I'm marking this resolved.

Like I said, I am open to arguments about how to handle different dependencies and their compatibility guarantees, but it's a nuanced and complex topic, so I'm not rashly making any change.

**Comment by ssbarnea** [comment]
@sirosen Thanks for bumping it. I mentioned the approach we took was only a suggestion based on our experience maintaining both libraries and applications.

I think we both know well at least the theory between library vs application dependencies. Still the reality is that the "app" mode, does apply more for end of the line projects, and not really all tools.

I can give you an endless list examples of applications which do not pin dependencies: ansible-lint, ansible, pip-tools. In fact it is very hard to find those that do pin them to exact versions. About 5 years ago pinning was more popular but it decreased and I hardly find one still doing them, even if they are considered applications.

Ceiling is perfecting fine, pinning the the problem. There are also other implication on pinning, repackaging downstream would likely break due to missing deps of conflicts. Downstream repackaging as in building rpms, debs or similar. That is not the case for us, but considering the generic purpose of check-jsonschema, I could see how some users might want to do it.

We discovered the `lock` extra as a decent workaround about one or two years ago. It was suggested by two different people involved with python packaging on their discourse, so is not really like to break in the future.

Again, not trying to push for any chance, just sharing my thoughts, hoping that you might find them useful.

**Comment by ssbarnea** [comment]
During last week, ruamel-yaml made 3 releases and each of them does impact rewriting support, which means that I had to bump it in ansible-lint (or the test suite will fail running). Current version is `0.17.24`... so probable you have a WTF moment while reading my comment.

I found a workaround for my use of python-jsonschema as now I am using it in a special testing pipeline which is using a separated virtualenv, at it would not be possible to install it alongside ansible-lint due to dependency conflict. Only by chance, this pipeline is implemented in nodejs instead of python, or otherwise I might end-up with similar conflicts.

I wish you a good weekend!
